<h1>Fernando's New Business</h1>
<h2> Description</h2>

Fernando is a mexican salesman, he has been working at a tec company for the last 30 years. He is finally retiring and wants to start a business with his savings. He doesn't know what kind of business he wants to run but he knows where he wants to open it, in Mexico City's wealthiest borough: Cuauhtemoc.

Fernando asked a Data Scientist for help. He wants him to find using data what kind of business to open and where he can open it. He wants little or no competition so he can earn big amounts of money so the data scientist will have to find the less common tipes of venues/businesses in each neighborhood.

<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li><a href="#Section_1"> Gathering the Data </a></li>
     <li><a href="#Section_2">Preparing the Data</a> </li>
     <li><a href="#Section_3">Obtaining the Venues</a></li>
     <li><a href="#Section_4">Finding out the least type of Venue per Neighborhood</a></li>
     <li><a href="#Section_5">Finding out the Where to Open the Business </a></li>
</div>

<hr>

<h2 id="section1"> Gathering the Data </h2>

The data scientist will get the neighborhood info from Foursquare, but for that he needs the list of neighborhoods inside the Cuauhtemoc borough. 

Fortunately, Mexico City's has a public database about the list of neighborhoods in the city. The information is puclic and esay to download in a csv file.

First import the necessary libraries for the project.

In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import numpy as np
import requests
import matplotlib.pyplot as plt
import matplotlib.cm as cm 
import matplotlib.colors as colors
import json
from pandas.io.json import json_normalize
from sklearn.cluster import KMeans
import folium

Then load the csv file and save it as a dataframe and translate the columns names.

The data is located in the following webpage.
https://datos.cdmx.gob.mx/dataset/coloniascdmx/resource/03368e1e-f05e-4bea-ac17-58fc650f6fee

In [2]:
mex_neigh = pd.read_csv("coloniascdmx.csv")
mex_neigh.rename(columns = {'nombre':'Neighborhood','alcaldia':'Borough'},inplace=True)
mex_neigh.head()

id                             Neighborhood  entidad  \
0   0                     LOMAS DE CHAPULTEPEC      9.0   
1   1  LOMAS DE REFORMA (LOMAS DE CHAPULTEPEC)      9.0   
2   2                     DEL BOSQUE (POLANCO)      9.0   
3   3               PEDREGAL DE SANTA URSULA I      9.0   
4   4                                 AJUSCO I      9.0   

                   geo_point_2d  \
0  19.4228411174,-99.2157935754   
1  19.4106158914,-99.2262487268   
2  19.4342189235,-99.2094037513   
3   19.314862237,-99.1477954505   
4   19.324571116,-99.1561602234   

                                           geo_shape  cve_alc         Borough  \
0  {"type": "Polygon", "coordinates": [[[-99.2201...       16  MIGUEL HIDALGO   
1  {"type": "Polygon", "coordinates": [[[-99.2296...       16  MIGUEL HIDALGO   
2  {"type": "Polygon", "coordinates": [[[-99.2082...       16  MIGUEL HIDALGO   
3  {"type": "Polygon", "coordinates": [[[-99.1458...        3        COYOACAN   
4  {"type": "Polygon", "coordinates": [[[-99.1585...        3        COYOACAN   

  cve_col                                  secc_com  \
0  16-042  4924, 4931, 4932, 4935, 4936, 4940, 4987   
1  16-044                                      4963   
2  16-026                                       NaN   
3  03-135                   433, 500, 431, 513, 501   
4  03-128        376, 377, 378, 379, 404, 493,  498   

                       secc_par  
0  4923, 4937, 4938, 4939, 4942  
1                          4964  
2                    4918, 4919  
3       424, 425, 426, 430, 499  
4                           374

The mex_neigh dataframe has all the neighborhoods from the city. The column Borough shows the neighborhood's borough and the geo_point_2d shows each neighborhood's coordinates. The rest of the columns are not useful so we can drop them.

<h2 id="section2"> Preparing the Data </h2>

The only thing that we need on this projects are the list of neighborhoods and their coordinates so the first step is to drop the columns that not useful for us.

In [3]:
mex_neigh.drop(columns=['id','entidad','geo_shape','cve_alc','cve_col','secc_com','secc_par'],axis=1,inplace=True)
mex_neigh.head()

Neighborhood                  geo_point_2d  \
0                     LOMAS DE CHAPULTEPEC  19.4228411174,-99.2157935754   
1  LOMAS DE REFORMA (LOMAS DE CHAPULTEPEC)  19.4106158914,-99.2262487268   
2                     DEL BOSQUE (POLANCO)  19.4342189235,-99.2094037513   
3               PEDREGAL DE SANTA URSULA I   19.314862237,-99.1477954505   
4                                 AJUSCO I   19.324571116,-99.1561602234   

          Borough  
0  MIGUEL HIDALGO  
1  MIGUEL HIDALGO  
2  MIGUEL HIDALGO  
3        COYOACAN  
4        COYOACAN

The column 'geo_point_2d' has the coordinates of each neighborhood. The next step is to split the coordinates and save it in Latitude and Longitude coordinates and drop the geo_point_2d column.

In [4]:
mex_split = mex_neigh['geo_point_2d'].str.split(',',expand=True)
mex_neigh['Latitude']=mex_split[0]
mex_neigh['Longitude']=mex_split[1]
mex_neigh.drop(columns=['geo_point_2d'],axis=1,inplace=True)
mex_neigh.head()

Neighborhood         Borough       Latitude  \
0                     LOMAS DE CHAPULTEPEC  MIGUEL HIDALGO  19.4228411174   
1  LOMAS DE REFORMA (LOMAS DE CHAPULTEPEC)  MIGUEL HIDALGO  19.4106158914   
2                     DEL BOSQUE (POLANCO)  MIGUEL HIDALGO  19.4342189235   
3               PEDREGAL DE SANTA URSULA I        COYOACAN   19.314862237   
4                                 AJUSCO I        COYOACAN   19.324571116   

        Longitude  
0  -99.2157935754  
1  -99.2262487268  
2  -99.2094037513  
3  -99.1477954505  
4  -99.1561602234

Fernando wants to open his bussiness in the Cuauhtemoc borough so let's create a new dataframe called cuau_neigh that has all the neighborhoods from the Cuauhtemoc borough.

In [5]:
mex_cuau = mex_neigh[mex_neigh['Borough']=='CUAUHTEMOC'].reset_index(drop=True)
mex_cuau.head()

Neighborhood     Borough       Latitude       Longitude
0                      TABACALERA  CUAUHTEMOC  19.4357759781  -99.1539492806
1                      CENTRO VII  CUAUHTEMOC  19.4302248036  -99.1281413675
2                      GUERRERO I  CUAUHTEMOC  19.4490761845  -99.1437494279
3  NONOALCO-TLATELOLCO (U HAB) II  CUAUHTEMOC  19.4533147946  -99.1417694775
4                          JUAREZ  CUAUHTEMOC  19.4270038256  -99.1616054122

Check if there are any non denifed values in the Latitude and Longitude columns.

In [6]:
mex_cuau[mex_cuau.Latitude.isnull()]

Neighborhood     Borough Latitude Longitude
13         MAZA  CUAUHTEMOC      NaN       NaN

The zip code for the Maza Neighborhood is 06270. The pgeocode library finds the location of a neighborhood based on its zip code. Find the zip code for the Maza neighoborhood and replace the unknown coordinates.

In [7]:
import pgeocode
nomi = pgeocode.Nominatim('mx')
Maza = nomi.query_postal_code("06270")
mex_cuau['Latitude']=mex_cuau['Latitude'].replace(np.nan,Maza.latitude)
mex_cuau['Longitude'] =mex_cuau['Longitude'].replace(np.nan,Maza.longitude)
mex_cuau

Neighborhood     Borough       Latitude       Longitude
0                        TABACALERA  CUAUHTEMOC  19.4357759781  -99.1539492806
1                        CENTRO VII  CUAUHTEMOC  19.4302248036  -99.1281413675
2                        GUERRERO I  CUAUHTEMOC  19.4490761845  -99.1437494279
3    NONOALCO-TLATELOLCO (U HAB) II  CUAUHTEMOC  19.4533147946  -99.1417694775
4                            JUAREZ  CUAUHTEMOC  19.4270038256  -99.1616054122
5               SANTA MARIA (U HAB)  CUAUHTEMOC  19.4564342667   -99.157053889
6                         CENTRO II  CUAUHTEMOC  19.4398500953  -99.1285178964
7                      ROMA NORTE I  CUAUHTEMOC  19.4194185761  -99.1691619817
8                         CENTRO IV  CUAUHTEMOC  19.4336362466  -99.1360300552
9                        ROMA SUR I  CUAUHTEMOC  19.4088498024  -99.1613175937
10                     SAN RAFAEL I  CUAUHTEMOC  19.4371086313  -99.1594601774
11                       CUAUHTEMOC  CUAUHTEMOC  19.4299776806  -99.1686722574
12                       DOCTORES I  CUAUHTEMOC  19.4239554155  -99.1472917929
13                             MAZA  CUAUHTEMOC        19.4547        -99.1279
14                          CONDESA  CUAUHTEMOC  19.4147504858  -99.1762103973
15                      VALLE GOMEZ  CUAUHTEMOC  19.4588759476  -99.1250261379
16                     BUENOS AIRES  CUAUHTEMOC  19.4071966156  -99.1446313482
17                      MORELOS III  CUAUHTEMOC  19.4453774475  -99.1333320245
18                         OBRERA I  CUAUHTEMOC  19.4195042636  -99.1387615628
19                       OBRERA III  CUAUHTEMOC  19.4107565162  -99.1417666594
20                        OBRERA II  CUAUHTEMOC  19.4151731171  -99.1392780039
21                    PERALVILLO II  CUAUHTEMOC  19.4607093703  -99.1306529543
22                        OBRERA IV  CUAUHTEMOC  19.4104067222  -99.1380145408
23         SANTA MARIA LA RIBERA IV  CUAUHTEMOC  19.4444233884  -99.1594591836
24                    SAN RAFAEL II  CUAUHTEMOC  19.4385711071  -99.1646939016
25                          ATLAMPA  CUAUHTEMOC  19.4586445841  -99.1563686743
26                  ASTURIAS (AMPL)  CUAUHTEMOC  19.4071962945  -99.1310629695
27                       CENTRO III  CUAUHTEMOC  19.4367707124  -99.1284523074
28    NONOALCO-TLATELOLCO (U HAB) I  CUAUHTEMOC  19.4549126344  -99.1480851254
29                      DOCTORES IV  CUAUHTEMOC  19.4119151507  -99.1518458397
30          SANTA MARIA INSURGENTES  CUAUHTEMOC  19.4616940155  -99.1487845392
31          SANTA MARIA LA RIBERA I  CUAUHTEMOC  19.4529587649  -99.1581421108
32                  PAULINO NAVARRO  CUAUHTEMOC  19.4117075418  -99.1299500041
33                          ALGARIN  CUAUHTEMOC  19.4055862513  -99.1402683369
34                     GUERRERO III  CUAUHTEMOC  19.4421924076  -99.1460093369
35                     BUENAVISTA I  CUAUHTEMOC  19.4496135645  -99.1495963852
36                      GUERRERO IV  CUAUHTEMOC  19.4404218952  -99.1422140732
37  NONOALCO-TLATELOLCO (U HAB) III  CUAUHTEMOC   19.451696421  -99.1353568265
38                      DOCTORES II  CUAUHTEMOC  19.4198737194  -99.1508495556
39                       DOCTORES V  CUAUHTEMOC  19.4103772864  -99.1467901424
40              SAN SIMON TOLNAHUAC  CUAUHTEMOC  19.4595317038  -99.1431802383
41                         ASTURIAS  CUAUHTEMOC  19.4046308544  -99.1312401576
42                    BUENAVISTA II  CUAUHTEMOC  19.4421892858  -99.1513520907
43                        CENTRO VI  CUAUHTEMOC  19.4247184898  -99.1346482655
44                      CENTRO VIII  CUAUHTEMOC  19.4315052317  -99.1463947881
45                     PERALVILLO I  CUAUHTEMOC  19.4627292503  -99.1374326001
46                    ROMA NORTE II  CUAUHTEMOC   19.421384627  -99.1588454263
47                         TRANSITO  CUAUHTEMOC  19.4178611105  -99.1315091546
48                HIPODROMO CONDESA  CUAUHTEMOC  19.4093352026  -99.1794803853
49       EX HIPODROMO DE PERALVILLO  CUAUHTEMOC  19.4566624075  -99.13477

Using geopy we can verify that all the neighborhoods on the dataframe are located on the Cuauhtemoc borough.

In [8]:
from geopy.geocoders import Nominatim
# create map
address = 'Centro, Mexico City'

geolocator = Nominatim(user_agent="mex_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of the borough Cuauhtemoc in Mexico City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of the borough Cuauhtemoc in Mexico City are 19.4065152, -99.1550183.


In [9]:
import folium
# create map of Mxico City using latitude and longitude values
map_mex = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(mex_cuau['Latitude'], mex_cuau['Longitude'], mex_cuau['Borough'], mex_cuau['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_mex)  
    
map_mex

With this data, the data scientist now can find the venues using foursquare.

<h2 id="section3"> Obtaining the Venues </h2>

The next step is to find out what kind of venues are located in the Cuahtemoc borough and where exactly are they located. In order to know that we can use the Foursquare API. First we have to geo our Foursquare credentials 

In [10]:
CLIENT_ID = 'PWJUI1C335YAEP5EZ2U0PR1K4AV5Z51BFC1LIOG5OGFCUXGC' # your Foursquare ID
CLIENT_SECRET = 'PM2FU40TUSHEIZ20ZOVOGC3GY4M2BK3SAJZ5EKNOEUV05YUV' # your Foursquare Secret
ACCESS_TOKEN = 'FEXV1SH3YRZIGPV1C3QA2KB21XRF510AALHTRBAHMWFS0L5P' # your FourSquare Access Token
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: PWJUI1C335YAEP5EZ2U0PR1K4AV5Z51BFC1LIOG5OGFCUXGC
CLIENT_SECRET:PM2FU40TUSHEIZ20ZOVOGC3GY4M2BK3SAJZ5EKNOEUV05YUV


With the credentials ready, we must find all the venues inside the Cuauhtemoc Borough, I will use the getNearbyvenues function that was defined in this course. Also, I need to impor the requests library to get the venues.

In [11]:
import requests
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
#        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(\
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

I will get the venues for each neighboorhod within the Cuauhtemoc borough and its coordinates and save them in a dataframe called cuau_venues.

In [12]:
cuau_venues = getNearbyVenues(mex_cuau['Neighborhood'],
                              mex_cuau['Latitude'],
                              mex_cuau['Longitude'])
cuau_venues.head()
print(cuau_venues.shape)
print("The are "+str(len(cuau_venues["Venue Category"].unique())) +" venue categories.")

(1726, 7)
The are 208 venue categories.


In conclusion, there are 1726 venues and 208 types of venue in the Cuahtemoc borough. 

<h2 id="section4"> Finding out the least type of Venue per Neighborhood </h2>

To help Fernando to decide what kind of bussiness he wants to open, is convenienly to look what types of venues are more common in each neigboorhood. So the first step is to find out the category of each venue. We can use the get_dummies function on Venue_Category column.

In [13]:
cuau_one_hot = pd.get_dummies(cuau_venues[["Venue Category"]])
cuau_one_hot["Neighborhood"] = cuau_venues["Neighborhood"]
# move neighborhood column to the first column
fixed_columns = [cuau_one_hot.columns[-1]] + list(cuau_one_hot.columns[:-1])
cuau_one_hot = cuau_one_hot[fixed_columns]
cuau_one_hot.head()

Neighborhood  Venue Category_Advertising Agency  \
0   TABACALERA                                  0   
1   TABACALERA                                  0   
2   TABACALERA                                  0   
3   TABACALERA                                  0   
4   TABACALERA                                  0   

   Venue Category_American Restaurant  Venue Category_Antique Shop  \
0                                   0                            0   
1                                   0                            0   
2                                   0                            0   
3                                   0                            0   
4                                   0                            0   

   Venue Category_Arcade  Venue Category_Argentinian Restaurant  \
0                      0                                      0   
1                      0                                      0   
2                      0                                      0   
3                      0                                      0   
4                      0                                      0   

   Venue Category_Art Gallery  Venue Category_Art Museum  \
0                           0                          0   
1                           0                          0   
2                           0                          0   
3                           0                          0   
4                           0                          0   

   Venue Category_Arts & Crafts Store  Venue Category_Arts & Entertainment  \
0                                   0                                    0   
1                                   0                                    0   
2                                   0                                    0   
3                                   0                                    0   
4                                   0                                    0   

   Venue Category_Asian Restaurant  Venue Category_Athletics & Sports  \
0                                0                                  0   
1                                0                                  0   
2                                0                                  0   
3                                0                                  0   
4                                0                                  0   

   Venue Category_Auto Garage  Venue Category_Auto Workshop  \
0                           0                             0   
1                           0                             0   
2                           0                             0   
3                           0                             0   
4                           0                             0   

   Venue Category_BBQ Joint  Venue Category_Bakery  Venue Category_Bar  \
0                         0                      0                   0   
1                         0                      0                   0   
2                         0                      0                   0   
3                         0                      0                   0   
4                         0                      0                   0   

   Venue Category_Basketball Court  Venue Category_Bed & Breakfast  \
0                                0                               0   
1                                0                               0   
2                                0                               0   
3                                0                               0   
4                                0                               0   

   Venue Category_Beer Bar  Venue Category_Beer Garden  \
0                        0                           0   
1                        0                           0   
2                        0                           0   
3                        0                           0   
4                        0                           0 

Now, let's group the dataframe by neighborhood and take the mean of the frequency. 

In [14]:
cuau_group = cuau_one_hot.groupby("Neighborhood").mean().reset_index()
print(cuau_group.shape)
cuau_group.head()

(64, 209)


Neighborhood  Venue Category_Advertising Agency  \
0          ALGARIN                           0.043478   
1         ASTURIAS                           0.000000   
2  ASTURIAS (AMPL)                           0.000000   
3          ATLAMPA                           0.000000   
4     BUENAVISTA I                           0.000000   

   Venue Category_American Restaurant  Venue Category_Antique Shop  \
0                            0.000000                          0.0   
1                            0.000000                          0.0   
2                            0.000000                          0.0   
3                            0.000000                          0.0   
4                            0.033333                          0.0   

   Venue Category_Arcade  Venue Category_Argentinian Restaurant  \
0               0.000000                               0.000000   
1               0.000000                               0.055556   
2               0.000000                               0.047619   
3               0.000000                               0.000000   
4               0.033333                               0.000000   

   Venue Category_Art Gallery  Venue Category_Art Museum  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         0.0                        0.0   
4                         0.0                        0.0   

   Venue Category_Arts & Crafts Store  Venue Category_Arts & Entertainment  \
0                                 0.0                                  0.0   
1                                 0.0                                  0.0   
2                                 0.0                                  0.0   
3                                 0.0                                  0.0   
4                                 0.0                                  0.0   

   Venue Category_Asian Restaurant  Venue Category_Athletics & Sports  \
0                              0.0                                0.0   
1                              0.0                                0.0   
2                              0.0                                0.0   
3                              0.0                                0.0   
4                              0.0                                0.0   

   Venue Category_Auto Garage  Venue Category_Auto Workshop  \
0                         0.0                           0.0   
1                         0.0                           0.0   
2                         0.0                           0.0   
3                         0.0                           0.0   
4                         0.0                           0.0   

   Venue Category_BBQ Joint  Venue Category_Bakery  Venue Category_Bar  \
0                  0.000000               0.086957            0.043478   
1                  0.055556               0.166667            0.055556   
2                  0.000000               0.000000            0.047619   
3                  0.000000               0.000000            0.000000   
4                  0.000000               0.000000            0.000000   

   Venue Category_Basketball Court  Venue Category_Bed & Breakfast  \
0                              0.0                        0.043478   
1                              0.0                        0.000000   
2                              0.0                        0.000000   
3                              0.0                        0.000000   
4                              0.0                        0.000000   

   Venue Category_Beer Bar  Venue Category_Beer Garden  \
0                      0.0                         0.0   
1                      0.0                         0.0   
2                      0.0                         0.0   
3                      0.0                         0.0   
4                      0.0         

Now, using the next two functions we can see the most and least common venues for each neighborhood.

In [15]:
def most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

def least_common_venues(row, num_least_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=True)
    
    return row_categories_sorted.index.values[0:num_least_venues]

First we are going to look at 20 the most common venues.

In [17]:
num_top_venues = 20

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns_top = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns_top.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns_top.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_common_venues_sorted = pd.DataFrame(columns=columns_top)
neighborhoods_common_venues_sorted['Neighborhood'] = cuau_group['Neighborhood']

for ind in np.arange(cuau_group.shape[0]):
    neighborhoods_common_venues_sorted.iloc[ind, 1:] = most_common_venues(cuau_group.iloc[ind, :], num_top_venues)
neighborhoods_common_venues_sorted.head()

Neighborhood              1st Most Common Venue  \
0          ALGARIN  Venue Category_Mexican Restaurant   
1         ASTURIAS              Venue Category_Bakery   
2  ASTURIAS (AMPL)          Venue Category_Taco Place   
3          ATLAMPA  Venue Category_Mexican Restaurant   
4     BUENAVISTA I       Venue Category_Movie Theater   

                   2nd Most Common Venue              3rd Most Common Venue  \
0              Venue Category_Taco Place              Venue Category_Bakery   
1      Venue Category_Mexican Restaurant      Venue Category_Ice Cream Shop   
2      Venue Category_Mexican Restaurant      Venue Category_Ice Cream Shop   
3            Venue Category_Optical Shop  Venue Category_Photography Studio   
4  Venue Category_Furniture / Home Store             Venue Category_Brewery   

              4th Most Common Venue              5th Most Common Venue  \
0     Venue Category_Clothing Store  Venue Category_Advertising Agency   
1          Venue Category_Juice Bar                Venue Category_Café   
2               Venue Category_Park               Venue Category_Diner   
3  Venue Category_Convenience Store            Venue Category_Pharmacy   
4     Venue Category_Ice Cream Shop            Venue Category_Boutique   

                 6th Most Common Venue              7th Most Common Venue  \
0            Venue Category_Print Shop         Venue Category_Coffee Shop   
1             Venue Category_Gift Shop                 Venue Category_Bar   
2                   Venue Category_Bar               Venue Category_Hotel   
3  Venue Category_Gym / Fitness Center           Venue Category_Nightclub   
4        Venue Category_Clothing Store  Venue Category_Frozen Yogurt Shop   

                 8th Most Common Venue           9th Most Common Venue  \
0            Venue Category_Food Truck  Venue Category_Bed & Breakfast   
1            Venue Category_Taco Place        Venue Category_BBQ Joint   
2             Venue Category_Gift Shop             Venue Category_Café   
3  Venue Category_Yucatecan Restaurant      Venue Category_Film Studio   
4   Venue Category_Sporting Goods Shop   Venue Category_Lingerie Store   

                10th Most Common Venue         11th Most Common Venue  \
0            Venue Category_Steakhouse             Venue Category_Bar   
1          Venue Category_Cupcake Shop    Venue Category_Burger Joint   
2           Venue Category_Flea Market      Venue Category_Restaurant   
3  Venue Category_Fast Food Restaurant  Venue Category_Farmers Market   
4         Venue Category_Shopping Mall          Venue Category_Garden   

                         12th Most Common Venue  \
0                        Venue Category_Brewery   
1  Venue Category_Vegetarian / Vegan Restaurant   
2                   Venue Category_Dessert Shop   
3             Venue Category_Falafel Restaurant   
4                  Venue Category_Luggage Store   

                         13th Most Common Venue  \
0           Venue Category_Gym / Fitness Center   
1                   Venue Category_Dessert Shop   
2  Venue Category_Vegetarian / Vegan Restaurant   
3                        Venue Category_Factory   
4                Venue Category_Bubble Tea Shop   

                  14th Most Common Venue  \
0               Venue Category_Gift Shop   
1  Venue Category_Argentinian Restaurant   
2        Venue Category_Video Game Store   
3              Venue Category_Eye Doctor   
4               Venue Category_Rock Club   

                  15th Most Common Venue        16th Most Common Venue  \
0             Venue Category_Escape Room    Venue Category_Film Studio   
1      Venue Category_Falafel Restaurant     Venue Category_Food Court   
2  Venue Category_Argentinian Restaurant  Venue Category_Burrito Place   
3                 Venue Category_Exhibit    Venue Category_Dry Cleaner   
4             Venue Category_Candy Store    Venue Category_Flea Market   

                17th Most Common Venue             18th Most Common Venu

For example, giving a quick look we can see that the most common venues are the Mexican restaurants, Taco places, bakeries and ice cream shops. Fernando doesn't want to much competition so he won't be interested in this types of venues. The next step is to look at the least common venues. 

In [19]:
num_least_venues = 20

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns_least = ['Neighborhood']
for ind in np.arange(num_least_venues):
    try:
        columns_least.append('{}{} Least Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns_least.append('{}th Least Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_least_venues_sorted = pd.DataFrame(columns=columns_least)
neighborhoods_least_venues_sorted['Neighborhood'] = cuau_group['Neighborhood']

for ind in np.arange(cuau_group.shape[0]):
    neighborhoods_least_venues_sorted.iloc[ind, 1:] = least_common_venues(cuau_group.iloc[ind, :], num_top_venues)
neighborhoods_least_venues_sorted.head()

Neighborhood             1st Least Common Venue  \
0          ALGARIN           Venue Category_Hotel Bar   
1         ASTURIAS  Venue Category_Advertising Agency   
2  ASTURIAS (AMPL)  Venue Category_Advertising Agency   
3          ATLAMPA  Venue Category_Advertising Agency   
4     BUENAVISTA I  Venue Category_Advertising Agency   

               2nd Least Common Venue          3rd Least Common Venue  \
0        Venue Category_Movie Theater           Venue Category_Museum   
1               Venue Category_Museum      Venue Category_Music Store   
2      Venue Category_Motorcycle Shop    Venue Category_Movie Theater   
3      Venue Category_Motorcycle Shop    Venue Category_Movie Theater   
4  Venue Category_Monument / Landmark  Venue Category_Motorcycle Shop   

       4th Least Common Venue                  5th Least Common Venue  \
0  Venue Category_Music Store              Venue Category_Music Venue   
1  Venue Category_Music Venue  Venue Category_New American Restaurant   
2       Venue Category_Museum              Venue Category_Music Store   
3       Venue Category_Museum              Venue Category_Music Store   
4       Venue Category_Museum              Venue Category_Music Store   

                   6th Least Common Venue  \
0  Venue Category_New American Restaurant   
1                Venue Category_Nightclub   
2              Venue Category_Music Venue   
3              Venue Category_Music Venue   
4              Venue Category_Music Venue   

                   7th Least Common Venue  \
0                Venue Category_Nightclub   
1               Venue Category_Non-Profit   
2  Venue Category_New American Restaurant   
3  Venue Category_New American Restaurant   
4  Venue Category_New American Restaurant   

                   8th Least Common Venue  \
0               Venue Category_Non-Profit   
1  Venue Category_North Indian Restaurant   
2                Venue Category_Nightclub   
3               Venue Category_Non-Profit   
4                Venue Category_Nightclub   

                   9th Least Common Venue  \
0  Venue Category_North Indian Restaurant   
1             Venue Category_Optical Shop   
2               Venue Category_Non-Profit   
3  Venue Category_North Indian Restaurant   
4               Venue Category_Non-Profit   

                  10th Least Common Venue  \
0             Venue Category_Optical Shop   
1     Venue Category_Other Great Outdoors   
2  Venue Category_North Indian Restaurant   
3     Venue Category_Other Great Outdoors   
4  Venue Category_North Indian Restaurant   

                        11th Least Common Venue  \
0           Venue Category_Other Great Outdoors   
1  Venue Category_Paper / Office Supplies Store   
2                   Venue Category_Optical Shop   
3  Venue Category_Paper / Office Supplies Store   
4             Venue Category_Miscellaneous Shop   

                        12th Least Common Venue  \
0  Venue Category_Paper / Office Supplies Store   
1                           Venue Category_Park   
2           Venue Category_Other Great Outdoors   
3                           Venue Category_Park   
4           Venue Category_Other Great Outdoors   

                        13th Least Common Venue  \
0                           Venue Category_Park   
1               Venue Category_Pedestrian Plaza   
2  Venue Category_Paper / Office Supplies Store   
3               Venue Category_Pedestrian Plaza   
4                           Venue Category_Park   

                14th Least Common Venue               15th Least Common Venue  \
0       Venue Category_Pedestrian Plaza  Venue Category_Performing Arts Venue   
1  Venue Category_Performing Arts Venue           Venue Category_Perfume Shop   
2       Venue Category_Pedestrian Plaza  Venue Category_Performing Arts Venue   
3  Venue Category_Performing Arts Venue           Venue Category_Perfume Shop   
4       Venue Category_Pedestrian Plaza  Venue Category_Performing Arts Venue   

              16th Lea

Doing the same as the last example, Fernando could decide into opening an advertising agecency, a movie theater, a music store or a new american restaurant and he could have little or no competition. This is the type of business that Fernando is looking.

<h2 id="section5"> Finding out the Where to Open the Business </h2>

Let's assume that Fernando has decided for a business to open thanks for the data we gave him. He went to look for a place to open his venue but he finds out that the rent is way to expensive or the space is to small, etc. Fernando has already made his mind and doesn't want to look for a new type of business, what can we do for him?

Well, using ML we can make clusters of neighborhood with similar properties, like having similar least common venues. So we are going to use k-means clustering to find similar neighborhoods where Fernando can look around for a place to rent and still have little to no competition. In this prohect we are going to make 15 clusters.

In [20]:
from sklearn.cluster import KMeans
kclusters = 15

cuau_cluster = cuau_group.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(cuau_cluster)

We have our cluster so now we are going to update our least common venues dataframe.

In [21]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

cuau_final = mex_cuau
cuau_final = cuau_final.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
cuau_final 

Neighborhood     Borough       Latitude  \
0                        TABACALERA  CUAUHTEMOC  19.4357759781   
1                        CENTRO VII  CUAUHTEMOC  19.4302248036   
2                        GUERRERO I  CUAUHTEMOC  19.4490761845   
3    NONOALCO-TLATELOLCO (U HAB) II  CUAUHTEMOC  19.4533147946   
4                            JUAREZ  CUAUHTEMOC  19.4270038256   
5               SANTA MARIA (U HAB)  CUAUHTEMOC  19.4564342667   
6                         CENTRO II  CUAUHTEMOC  19.4398500953   
7                      ROMA NORTE I  CUAUHTEMOC  19.4194185761   
8                         CENTRO IV  CUAUHTEMOC  19.4336362466   
9                        ROMA SUR I  CUAUHTEMOC  19.4088498024   
10                     SAN RAFAEL I  CUAUHTEMOC  19.4371086313   
11                       CUAUHTEMOC  CUAUHTEMOC  19.4299776806   
12                       DOCTORES I  CUAUHTEMOC  19.4239554155   
13                             MAZA  CUAUHTEMOC        19.4547   
14                          CONDESA  CUAUHTEMOC  19.4147504858   
15                      VALLE GOMEZ  CUAUHTEMOC  19.4588759476   
16                     BUENOS AIRES  CUAUHTEMOC  19.4071966156   
17                      MORELOS III  CUAUHTEMOC  19.4453774475   
18                         OBRERA I  CUAUHTEMOC  19.4195042636   
19                       OBRERA III  CUAUHTEMOC  19.4107565162   
20                        OBRERA II  CUAUHTEMOC  19.4151731171   
21                    PERALVILLO II  CUAUHTEMOC  19.4607093703   
22                        OBRERA IV  CUAUHTEMOC  19.4104067222   
23         SANTA MARIA LA RIBERA IV  CUAUHTEMOC  19.4444233884   
24                    SAN RAFAEL II  CUAUHTEMOC  19.4385711071   
25                          ATLAMPA  CUAUHTEMOC  19.4586445841   
26                  ASTURIAS (AMPL)  CUAUHTEMOC  19.4071962945   
27                       CENTRO III  CUAUHTEMOC  19.4367707124   
28    NONOALCO-TLATELOLCO (U HAB) I  CUAUHTEMOC  19.4549126344   
29                      DOCTORES IV  CUAUHTEMOC  19.4119151507   
30          SANTA MARIA INSURGENTES  CUAUHTEMOC  19.4616940155   
31          SANTA MARIA LA RIBERA I  CUAUHTEMOC  19.4529587649   
32                  PAULINO NAVARRO  CUAUHTEMOC  19.4117075418   
33                          ALGARIN  CUAUHTEMOC  19.4055862513   
34                     GUERRERO III  CUAUHTEMOC  19.4421924076   
35                     BUENAVISTA I  CUAUHTEMOC  19.4496135645   
36                      GUERRERO IV  CUAUHTEMOC  19.4404218952   
37  NONOALCO-TLATELOLCO (U HAB) III  CUAUHTEMOC   19.451696421   
38                      DOCTORES II  CUAUHTEMOC  19.4198737194   
39                       DOCTORES V  CUAUHTEMOC  19.4103772864   
40              SAN SIMON TOLNAHUAC  CUAUHTEMOC  19.4595317038   
41                         ASTURIAS  CUAUHTEMOC  19.4046308544   
42                    BUENAVISTA II  CUAUHTEMOC  19.4421892858   
43                        CENTRO VI  CUAUHTEMOC  19.4247184898   
44                      CENTRO VIII  CUAUHTEMOC  19.4315052317   
45                     PERALVILLO I  CUAUHTEMOC  19.4627292503   
46                    ROMA NORTE II  CUAUHTEMOC   19.421384627   
47                         TRANSITO  CUAUHTEMOC  19.4178611105   
48                HIPODROMO CONDESA  CUAUHTEMOC  19.4093352026   
49       EX HIPODROMO DE PERALVILLO  CUAUHTEMOC  19.4566624075   
50                         CENTRO I  CUAUHTEMOC   19.441997265   
51                      HIPODROMO I  CUAUHTEMOC  19.4121695524   
52                   ROMA NORTE III  CUAUHTEMOC  19.4147537971   
53                      ROMA SUR II  CUAUHTEMOC  19.4040153783   
54                      GUERRERO II  CUAUHTEMOC  19.4470270735   
55                     VISTA ALEGRE  CUAUHTEMOC  19.4101695171   
56                       MORELOS II  CUAUHTEMOC  19.4466164805   
57                  FELIPE PESCADOR  CUAUHTEMOC  19.4540778349   
58         SANTA MARIA LA RIBERA II  CUAUHTEMOC  19.4492196589   
59                        ESPERANZA  CUAUHTEMOC  19.4204440063 

For the final step, to make thinfgs easier for Fernando, we are going to show all the clister into the Cuahtemoc map so he can find the similar neighborghoods easier.

In [22]:
import matplotlib.pyplot as plt
import matplotlib.cm as cm 
import matplotlib.colors as colors
# create map
address = 'Centro,Mexico City'

geolocator = Nominatim(user_agent="mex_explorer")
location = geolocator.geocode(address)
lat = location.latitude
long = location.longitude
map_clusters = folium.Map(location=[lat, long], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(cuau_final['Latitude'], cuau_final['Longitude'], cuau_final['Neighborhood'], cuau_final['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

With all this information, Fernando have the tools to pick what kind of business he want to open and places where he can open it.